File: interviews/leetcode/leetcode-75/trie/208. Implement Prefix Trie/solution.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEnd = False

class Trie:

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str):
        node = self.root
        for ch in word:
            node.children.setdefault(ch, TrieNode())
            node = node.children[ch]
        node.isEnd = True

    def search(self, word: str) -> bool:
        node = self.root
        for ch in word:
            if ch not in node.children:
                return False
            node = node.children[ch]
        return node.isEnd

    def startsWith(self, prefix: str) -> bool:
        node = self.root
        for ch in prefix:
            if ch not in node.children:
                return False
            node = node.children[ch]
        return True

# Main function (commented out by default)
# def main():
#     trie = Trie()
#     trie.insert("apple")
#     print(trie.search("apple"))   # true
#     print(trie.search("app"))     # false
#     print(trie.startsWith("app")) # true
#     trie.insert("app")
#     print(trie.search("app"))     # true

File: interviews/leetcode/leetcode-75/trie/208. Implement Prefix Trie/solution.go
package main

// TrieNode represents a node in the Trie.
type TrieNode struct {
	children map[rune]*TrieNode
	isEnd    bool
}

// Trie represents the Trie data structure.
type Trie struct {
	root *TrieNode
}

// Constructor to create a new Trie.
func Constructor() Trie {
	return Trie{root: &TrieNode{children: make(map[rune]*TrieNode)}}
}

// Insert inserts a word into the Trie.
func (t *Trie) Insert(word string) {
	node := t.root
	for _, ch := range word {
		if _, exists := node.children[ch]; !exists {
			node.children[ch] = &TrieNode{children: make(map[rune]*TrieNode)}
		}
		node = node.children[ch]
	}
	node.isEnd = true
}

// Search checks if a word exists in the Trie.
func (t *Trie) Search(word string) bool {
	node := t.root
	for _, ch := range word {
		if _, exists := node.children[ch]; !exists {
			return false
		}
		node = node.children[ch]
	}
	return node.isEnd
}

// StartsWith checks if there is any word in the Trie that starts with the given prefix.
func (t *Trie) StartsWith(prefix string) bool {
	node := t.root
	for _, ch := range prefix {
		if _, exists := node.children[ch]; !exists {
			return false
		}
		node = node.children[ch]
	}
	return true
}

// Main function (commented out by default)
// func main() {
//     trie := Constructor()
//     trie.Insert("apple")
//     fmt.Println(trie.Search("apple"))   // true
//     fmt.Println(trie.Search("app"))     // false
//     fmt.Println(trie.StartsWith("app")) // true
//     trie.Insert("app")
//     fmt.Println(trie.Search("app"))     // true
// }
